Update docstrings on BackendV2Converter and convert_to_target

itoko opened the pull request:
A custom calibration defined in `backend.defaults()` was silently dropped during the conversion to a target using `convert_to_target` in a certain case. Consequently, a `BackendV2` object converted from a `BackendV1` object using `BackendV2Converter` may not contain a custom calibration added to `BackendV1.defaults()` accidentally.
This commit adds a docstring that clearly describes when a custom calibration in `defaults` may be dropped and it is not recommended to update a BackendV1 object before conversion to `convert_to_target` and `BackendV2Converter` to avoid such a case.

mtreinish reviewed (commented):
I'm a bit concerned about how noisy this will be. I've learned the hard way in the past that warnings like this which aren't necessarily user actionable are more of an annoyance than a help. But, I think in this case it's sufficiently uncommon that if you have a backend with a custom calibration in the `InstructionScheduleMap` you'd probably want to know that it's being dropped.

So that beign said  I'm fine to add this warning. If people start to complain about we can always remove it in the future. Before, we merge this though I think it would be good to add this details about custom calibrations to the docstrings of `convert_to_target` and `BackendV2Converter` too

nkanazawa1989 commented on qiskit/providers/backend_compat.py:
```diff
@@ -132,6 +133,14 @@ def convert_to_target(
                             target[inst][(qubit,)].calibration = calibration_entry
                     elif qargs in target[inst]:
                         target[inst][qargs].calibration = calibration_entry
+            if inst not in target and inst not in name_mapping:
+                warnings.warn(
+                    f"Custom calibration for '{inst}' defined in 'defaults' is dropped "
+                    "during the conversion to a target. If you don't want that, "
+                    "supply a custom_name_mapping or "
+                    "add the custom calibration to a target after building the target.",
+                    UserWarning,
+                )
```
```suggestion
                else:
                    warnings.warn(
                        f"Custom calibration for '{inst}' defined in 'defaults' is dropped "
                        "during the conversion to a target. If you don't want that, "
                        "supply a custom_name_mapping or "
                        "add the custom calibration to a target after building the target.",
                        UserWarning,
                    )
```
How does checking the `name_mapping` help to detect missing entry? Probably you can check [`calibration_entry.user_provided`](https://github.com/Qiskit/qiskit-terra/blob/58aceb59e9bc1e16f29b09402e1984b16d211da3/qiskit/pulse/calibration_entries.py#L100-L102) to selectively trigger the warning.

nkanazawa1989 replied:
Okey. Target entry cannot be generated without having `Instruction` mapping and this looks somewhat obvious. On the other hand, I think there is the case a user provided `custom_name_mapping` but calibration is missing because key name is not registered to the basis_gates. In this case instmap entry is also silently gone, but warning is not raise because it checks `name_mapping`.

nkanazawa1989 reviewed (commented):
I also think it's worth adding the warning especially for missing entries that user intentionally added to the inst map. Alternatively we can update converter logic to pick them.

itoko commented:
I've talked with Naoki and agreed that updating docstrings would be sufficient in this case considering use cases in qiskit-experiments. I've updated docstrings and removed the raise of warning. Could you take a final look? @mtreinish

mtreinish commented on qiskit/providers/backend_compat.py:
```diff
@@ -206,6 +217,21 @@ class BackendV2Converter(BackendV2):
     class should only be used if you need a :class:`~.BackendV2` and still need
     compatibility with :class:`~.BackendV1`.
+
+    Note that updating a ``BackendV1`` object before applying this converter is **not** recommended.
+    For example, in order to convert a ``BackendV1`` object with a customized
+    ``defaults().instruction_schedule_map``, which has a custom calibration for an operation,
+    the operation name must be in ``configuration().basis_gates`` and
+    ``name_mapping`` must be supplied for the operation.
+    Otherwise, the operation will be dropped in the resulting ``BackendV2`` object.
+
+    That suggests it is recommended to add custom calibrations **after** applying this converter
+    instead of updating ``BackendV1.defaults()`` in advance. For example::
+
+        backend_v2 = BackendV2Converter(backend_v1)
+        backend_v2.target.add_instruction(
+            custom_gate, {(0, 1): InstructionProperties(calibration=custom_sched)}
+        )
```
```suggestion
    When using custom calibrations (or other custom workflows) it is **not** recommended
    to mutate the ``BackendV1`` object before applying this converter. For example, in order to
    convert a ``BackendV1`` object with a customized ``defaults().instruction_schedule_map``,
    which has a custom calibration for an operation, the operation name must be in
     ``configuration().basis_gates`` and  ``name_mapping`` must be supplied for the operation.
    Otherwise, the operation will be dropped in the resulting ``BackendV2`` object.

    Instead it is typically better to add custom calibrations **after** applying this converter
    instead of updating ``BackendV1.defaults()`` in advance. For example::

        backend_v2 = BackendV2Converter(backend_v1)
        backend_v2.target.add_instruction(
            custom_gate, {(0, 1): InstructionProperties(calibration=custom_sched)}
        )
```

mtreinish reviewed (commented):
Overall this LGTM. Just a small tweak suggested to the wording on the `BackendV2Converter` docstring. If that's ok for you lets apply it and this should be good to merge.
